### PR TITLE
[Feature:Submission] Display dispatcher/graphics action list to students

### PR DESCRIPTION
**Pull request by emaicus**
This PR allows instructors to specify the field `publish_actions` at either the global or testcase level of a configuration for an assignment which uses dispatcher or graphics actions. When set to `true`, all actions for a testcase are displayed to the student on the assignment submission page.

### CODE REVIEW

**Comment by emaicus** [comment]
Ready for review. To review,
1. set `publish_actions` to true for tutorials 15 or 16.
2. Submit anything to the tutorial (score doesn't matter).
3. Verify that the `View test procedure` button is present and expand it.

File: site/public/js/submission-page.js [hunk 1]
```diff
@@ -1,5 +1,10 @@
-function showActions(form_id) {
-    var form = $("#" + form_id);
-    form.css("display", "block");
-    form.find('.form-body').scrollTop(0);
+
+function openActionsPopup(popup_css, server_css, element_id) {
+  var elem_html = "<link rel=\"stylesheet\" type=\"text/css\" href=\"" + server_css + "\" />";
+  elem_html += "<link rel=\"stylesheet\" type=\"text/css\" href=\"" + popup_css + "\" />";
+  elem_html += document.getElementById(element_id).innerHTML;
```
Comment (MasterOdin) [inline site/public/js/submission-page.js:5]: there should probably be some minimal padding around the left/right side of content as having it start at 0px is a bit ugly.

File: site/public/js/submission-page.js [hunk 2]
```diff
@@ -1,5 +1,10 @@
-function showActions(form_id) {
-    var form = $("#" + form_id);
-    form.css("display", "block");
-    form.find('.form-body').scrollTop(0);
+
+function openActionsPopup(popup_css, server_css, element_id) {
+  var elem_html = "<link rel=\"stylesheet\" type=\"text/css\" href=\"" + server_css + "\" />";
```
Comment (MasterOdin) [inline site/public/js/submission-page.js:3]:
```suggestion
  let elem_html = `<link rel="stylesheet" type="text/css" href="${server_css}" />`;
```

File: site/public/js/submission-page.js [hunk 3]
```diff
@@ -1,5 +1,10 @@
...
+  elem_html += "<link rel=\"stylesheet\" type=\"text/css\" href=\"" + popup_css + "\" />";
```
Comment (MasterOdin) [inline site/public/js/submission-page.js:4]:
```suggestion
  elem_html += `<link rel="stylesheet" type="text/css" href="${popup_css}" />`;
```

File: site/public/js/submission-page.js (hunk 2 above)
Comment (MasterOdin) [reply site/public/js/submission-page.js:3]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

**Comment by MasterOdin** [comment]
I feel like we should be numbering the actions of the test procedure, especially if there's a change that one could go past a single line.

File: site/app/models/gradeable/AutogradingTestcase.php [hunk 4]
```diff
@@ -36,6 +36,12 @@ class AutogradingTestcase extends AbstractModel {
     protected $view_testcase_message = true;
     /** @prop @var string */
     protected $testcase_label = '';
+    /** @prop @var array */
+    protected $dispatcher_actions = array();
+    /** @prop @var array */
+    protected $graphics_actions = array();
+    /** @prop @var boolean */
+    protected $publish_actions = false;
```
Comment (MasterOdin) [inline site/app/models/gradeable/AutogradingTestcase.php:44]: `@prop-read` accomplishes the same thing as adding the `set` functions that throw. See https://submitty.org/developer/developing_the_php_site/models.

Comment (MasterOdin) [reply site/app/models/gradeable/AutogradingTestcase.php:44]: Also, please add a `@method getDispatcherActions` et al. above the class as that is used by phpstan for type hinting wherever the functions are called.

File: site/app/views/AutoGradingView.php [hunk 5]
```diff
@@ -229,13 +229,19 @@ public function loadAutoChecks(\app\models\gradeable\GradedGradeable $graded_gra
...
             "display_version" => $version->getVersion(),
             "index" => $testcase->getTestcase()->getIndex(),
             "who" => $who,
-            "popup_css_file" => $popup_css_file
+            "popup_css_file" => $popup_css_file,
+            "server_css_file" => $server_css_file,
+            "testcase_name" => $testcase->getTestcase()->getName(),
+            "publish_actions" => $testcase->getTestcase()->isPublishActions(),
+            "dispatcher_actions" => $testcase->getTestcase()->getDispatcherActions(),
+            "graphics_actions" => $testcase->getTestcase()->getGraphicsActions()
```
Comment (MasterOdin) [inline site/app/views/AutoGradingView.php:244]: Just pass in `"testcase" => $testcase->getTestcase()` as a single object into Twig and then call the methods from within Twig.

**Comment by bmcutler** [comment]
@emaicus 
For consistent documentation (PR notes are release notes), 
we've decided to use light mode for all PR screenshots (including dark mode images is optional).
Please edit your top post to add light mode image (or replace).
Thanks!

File: site/public/js/submission-page.js [hunk 6]
```diff
@@ -0,0 +1,9 @@
+
+function openActionsPopup(popup_css, element_id) {
+  elem_html = `<link rel="stylesheet" type="text/css" href="${popup_css}" />`;
```
Comment (MasterOdin) [inline site/public/js/submission-page.js:3]:
```suggestion
  let elem_html = `<link rel="stylesheet" type="text/css" href="${popup_css}" />`;
```

**Pull request merged**